### ISSUE TITLE
Enhancement: Support for returning IDictionary<string, object> instead of ExpandoObject.

### ISSUE THREAD

**Issue by isaacabraham** [mikependon/RepoDB#537]
F# doesn't have "out of the box" support for dynamic typing (although it's something that you *could* add support for). Instead, it would be nice to be able to request an `IDictionary<String,Object>>` from a query e.g.

```fsharp
let data = conn.ExecuteQuery<IDictionary<string, obj>> "SELECT Col1, Col2 FROM Table" // data is type IEnumerable<IDictionary<string, obj>>

let stuff = data |> Seq.map(fun row -> row.["Col1"]) // get a list of all Col1 values
```

You can work around this with a one liner, but not everyone (and especially developers with no C# experience) will know how to do this:

```fsharp
let data = conn.ExecuteQuery "SELECT Col1, Col2 FROM Table" // data is type IEnumerable<obj>
let dataAsDict = data |> Seq.cast<IDictionary<string, obj>> // dataAsDict is type IEnumerable<IDictionary<string,obj>>

**Comment by isaacabraham** [comment]
As background, in F# you can implement the `?` operator to do any kind of dynamic lookup so you *could* do this...

```fsharp
let (?) (x:obj) (key) = (x :?> IDictionary<string, obj>).[key]
let data = conn.ExecuteQuery "SELECT Col1, Col2 FROM Table" // data is type IEnumerable<obj>
let stuff = data |> Seq.map(fun row -> row?Col1) // Note the use of "?" here.

**Comment by mikependon** [comment]
Just to clarify, the resultset of this is of type `IEnumerable<IDictionary<string, object>>` right? In which, all the mentioned columns types from the `SELECT` statement will be casted back to `System.Object`.

**Comment by isaacabraham** [comment]
Yes. When you say "all the mentioned columns types" I assume you mean the values in each row (not the column names!)?

And yes, that's what the Expando Object is underneath the hood. It just wraps it in a special interface that C# knows about to give it the dynamic capabilities.

**Comment by mikependon** [comment]
> Yes. When you say "all the mentioned columns types" I assume you mean the values in each row (not the column names!)?

Yes, the name = `System.String` and the value = `System.Object`.

**Comment by isaacabraham** [comment]
Right. I'm not sure what else it could really be? This is what the Expando is behind the scenes. The only other option is potentially to expose the "raw" ADO .NET DB type out so that you can safely ask it what type it is before casting from object to e.g. string etc.

**Comment by mikependon** [comment]
Cool! I will try to see if I can get the data type back as I already have the proper resolver for that. The only reason I ask is because ExpandoObject is only limited to IDictionary<string, object>.

**Comment by isaacabraham** [comment]
I think realistically it would probably need to be a <string, obj> collection because each row may have different types in across the columns.

**Comment by mikependon** [comment]
Atleast I can create a thing like this `IEnumerable<IDictionary<string, RowValue>>`.

```csharp
public class RowValue
{
	public Type DataType { get; set; }
	public object Value { get; set; } /* This is always object */
}
```

But I am also afraid to dictate to the user to bind their code to that `RowValue` object. So, for the sake of simplicity, let us agree with the `IDictionary<string, object>`, there I will start the implementation.

**Comment by mikependon** [comment]
The fix is available at version [RepoDb v1.12.0-beta3](https://www.nuget.org/packages/RepoDb/1.12.0-beta3) or above.

**Comment by mikependon** [comment]
@isaacabraham - we are closing this ticket now. Please do let us know if you still have questions. Thanks
